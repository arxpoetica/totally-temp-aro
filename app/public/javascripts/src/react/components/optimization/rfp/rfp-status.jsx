import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import reduxStore from '../../../../redux-store'
import wrapComponentWithProvider from '../../../common/provider-wrapped-component'
import RfpActions from './rfp-actions'

class RfpReportDownloadCell extends Component {
  constructor (props) {
    super(props)
    this.state = {
      selectedReportTypeId: this.props.reportDefinitions[0].reportData.id
    }
  }

  render () {
    const selectedReport = this.props.reportDefinitions.filter(report => report.reportData.id === this.state.selectedReportTypeId)[0]
    return <div className='d-flex'>
      <select
        className='form-control'
        style={{ marginTop: '1px' }}
        value={this.state.selectedReportTypeId}
        onChange={event => this.setState({ selectedReportTypeId: +event.target.value })}>
        {this.props.reportDefinitions.map(reportDefinition => (
          <option
            key={reportDefinition.reportData.id}
            value={reportDefinition.reportData.id}>
            {reportDefinition.reportData.displayName}
          </option>
        ))}
      </select>
      <div className='btn btn-group p-0'>
        {
          selectedReport.reportData.media_types.map(mediaType => {
            // "(new Date()).toISOString().split('T')[0]" will give "YYYY-MM-DD"
            // Note that we are doing (new Date(Date.now())) so that we can have deterministic tests (by replacing the Date.now() function when testing)
            const downloadFileName = `${(new Date(Date.now())).toISOString().split('T')[0]}_${selectedReport.reportData.name}.${mediaType}`
            const downloadUrl = selectedReport.href
              .replace('{planId}', this.props.planId)
              .replace('{mediaType}', mediaType)
              .replace('{userId}', this.props.userId)
            return <a
              key={mediaType}
              className='btn btn-light'
              style={{ whiteSpace: 'nowrap' }}
              href={`/service-download-file/${downloadFileName}${downloadUrl}`}
              download>
              <i className='fa fa-download' /> {mediaType}
            </a>
          })
        }
      </div>
    </div>
  }
}

RfpReportDownloadCell.propTypes = {
  planId: PropTypes.number,
  reportDefinitions: PropTypes.arrayOf(PropTypes.shape({
    reportData: PropTypes.shape({
      id: PropTypes.number,
      reportType: PropTypes.string,
      name: PropTypes.string,
      displayName: PropTypes.string,
      media_types: PropTypes.arrayOf(PropTypes.string)
    }),
    href: PropTypes.string
  })),
  userId: PropTypes.number
}

class RfpStatusRow extends Component {
  render () {
    const planStateToBadgeColor = {
      UNDEFINED: 'badge-danger',
      START_STATE: 'badge-dark',
      INITIALIZED: 'badge-dark',
      STARTED: 'badge-primary',
      COMPLETED: 'badge-success',
      CANCELED: 'badge-danger',
      FAILED: 'badge-danger'
    }
    const createdByActor = this.props.systemActors[this.props.createdById]
    const createdByName = createdByActor ? `${createdByActor.firstName} ${createdByActor.lastName}` : `UserID: ${this.props.createdById}`
    return <tr style={{ textAlign: 'center' }}>
      <td>{this.props.planId}</td>
      <td>{this.props.name}</td>
      <td>{createdByName}</td>
      <td>
        <div className={`badge ${planStateToBadgeColor[this.props.status]}`}>{this.props.status}</div>
      </td>
      <td>
        <RfpReportDownloadCell planId={this.props.planId} userId={this.props.userId} reportDefinitions={this.props.reportDefinitions} />
      </td>
    </tr>
  }
}

RfpStatusRow.propTypes = {
  planId: PropTypes.number,
  name: PropTypes.string,
  createdById: PropTypes.number,
  status: PropTypes.string,
  systemActors: PropTypes.object,
  reportDefinitions: PropTypes.arrayOf(PropTypes.shape({
    reportData: PropTypes.shape({
      id: PropTypes.number,
      reportType: PropTypes.string,
      name: PropTypes.string,
      displayName: PropTypes.string,
      media_types: PropTypes.arrayOf(PropTypes.string)
    }),
    href: PropTypes.string
  })),
  userId: PropTypes.number
}

export class RfpStatus extends Component {
  render () {
    return <div className='container pt-5'>
      <h2>RFP Plan Status</h2>
      <div className='row'>
        <div className='col-md-12'>
          <table className='table table-sm table-striped'>
            <thead className='thead-light'>
              <tr style={{ textAlign: 'center' }}>
                <th>ID</th>
                <th>Name</th>
                <th>Created by</th>
                <th>Status</th>
                <th style={{ width: '400px' }}>Reports</th>
              </tr>
            </thead>
            <tbody>
              {
                this.props.isLoadingRfpPlans
                  ? this.renderLoadingIconRow()
                  : this.renderRfpPlanRows()
              }
            </tbody>
          </table>
          {this.renderPagination()}
        </div>
      </div>
    </div>
  }

  renderLoadingIconRow () {
    return <tr>
      <td colSpan={5} className='p-5 text-center'>
        <div className='fa fa-5x fa-spin fa-spinner mb-4' />
        <h4>Loading RFP Plans...</h4>
      </td>
    </tr>
  }

  renderRfpPlanRows () {
    const plansToRender = this.props.rfpPlans.slice(this.props.planListOffset, this.props.planListOffset + this.props.planListLimit)
    return plansToRender.map(rfpPlan => (
      <RfpStatusRow
        key={rfpPlan.id}
        planId={rfpPlan.id}
        name={rfpPlan.name}
        createdById={rfpPlan.createdBy}
        status={rfpPlan.planState}
        systemActors={this.props.systemActors}
        reportDefinitions={this.props.rfpReportDefinitions}
        userId={this.props.userId}
      />
    ))
  }

  renderPagination () {
    const numPlans = this.props.rfpPlans.length
    if (numPlans === 0) {
      return null
    }
    const activePageNumber = Math.round(this.props.planListOffset / this.props.planListLimit) + 1
    const numPages = Math.ceil(numPlans / this.props.planListLimit)
    const NUM_PAGES_TO_SHOW = 4 // Number of pages to show before/after the active page
    const startPage = Math.max(activePageNumber - NUM_PAGES_TO_SHOW, 1)
    const endPage = Math.min(activePageNumber + NUM_PAGES_TO_SHOW, numPages)
    const planListLimit = this.props.planListLimit
    const offsetOnClick = pageToLoad => (pageToLoad - 1) * planListLimit
    var pageBlocks = []
    if (activePageNumber > 1) {
      pageBlocks.push(<li key={'Previous'} className='page-item'>
        <a className='page-link' href='#' onClick={event => this.props.setPlanListOffset(offsetOnClick(activePageNumber - 1))}>Prev</a>
      </li>)
    }
    for (var iPage = startPage; iPage <= endPage; ++iPage) {
      const liClassName = 'page-item' + (iPage === activePageNumber ? ' active' : '')
      const copyIPage = iPage
      pageBlocks.push(<li key={iPage} className={liClassName}>
        <a className='page-link' href='#' onClick={event => this.props.setPlanListOffset(offsetOnClick(copyIPage))}>{iPage}</a>
      </li>)
    }
    if (activePageNumber !== numPages) {
      pageBlocks.push(<li key={'Next'} className='page-item'>
        <a className='page-link' href='#' onClick={event => this.props.setPlanListOffset(offsetOnClick(activePageNumber + 1))}>Next</a>
      </li>)
    }
    return <nav aria-label='RFP plan pagination'>
      <ul className='pagination justify-content-center'>
        {pageBlocks}
      </ul>
    </nav>
  }

  componentDidMount () {
    this.props.loadRfpPlans(this.props.userId)
  }

  componentWillUnmount () {
    this.props.clearRfpPlans()
  }
}

RfpStatus.propTypes = {
  rfpPlans: PropTypes.array,
  rfpReportDefinitions: PropTypes.arrayOf(PropTypes.shape({
    reportData: PropTypes.shape({
      id: PropTypes.number,
      reportType: PropTypes.string,
      name: PropTypes.string,
      displayName: PropTypes.string,
      media_types: PropTypes.arrayOf(PropTypes.string)
    }),
    href: PropTypes.string
  })),
  isLoadingRfpPlans: PropTypes.bool,
  planListOffset: PropTypes.number,
  planListLimit: PropTypes.number,
  systemActors: PropTypes.object,
  userId: PropTypes.number
}

const mapStateToProps = state => ({
  rfpPlans: state.optimization.rfp.rfpPlans,
  rfpReportDefinitions: state.optimization.rfp.rfpReportDefinitions,
  isLoadingRfpPlans: state.optimization.rfp.isLoadingRfpPlans,
  planListOffset: state.optimization.rfp.planListOffset,
  planListLimit: state.optimization.rfp.planListLimit,
  systemActors: state.user.systemActors,
  userId: state.user.loggedInUser && state.user.loggedInUser.id
})

const mapDispatchToProps = dispatch => ({
  clearRfpPlans: () => dispatch(RfpActions.clearRfpPlans()),
  loadRfpPlans: userId => dispatch(RfpActions.loadRfpPlans(userId)),
  setPlanListOffset: planListOffset => dispatch(RfpActions.setPlanListOffset(planListOffset))
})

const RfpStatusComponent = wrapComponentWithProvider(reduxStore, RfpStatus, mapStateToProps, mapDispatchToProps)
export default RfpStatusComponent
