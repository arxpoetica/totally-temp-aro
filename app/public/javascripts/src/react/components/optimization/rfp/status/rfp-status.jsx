import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import reduxStore from '../../../../../redux-store'
import wrapComponentWithProvider from '../../../../common/provider-wrapped-component'
import RfpActions from '../rfp-actions'
import RfpPlanList from './rfp-plan-list.jsx'
import RfpSubmitter from './rfp-submitter.jsx'
import RfpTemplateManager from './rfp-template-manager.jsx'

export class RfpStatus extends Component {
  render () {
    return <div className='container pt-5 pb-5'>
      <h2>RFP Plans</h2>
      <ul className='nav nav-tabs mb-3'>
        {
          this.props.tabs.map(tab => (
            <li key={tab.id} className='nav-item'>
              <a
                id={`rfpStatusTab_${tab.id}`}
                className={`nav-link ${tab.id === this.props.selectedTabId ? 'active' : ''}`}
                href='#'
                onClick={() => this.props.setSelectedTabId(tab.id)}
              >
                {tab.description}
              </a>
            </li>
          ))
        }
      </ul>
      {this.renderActiveComponent()}
    </div>
  }

  renderActiveComponent () {
    switch (this.props.selectedTabId) {
      case 'LIST_PLANS':
        return <RfpPlanList />

      case 'SUBMIT_RFP':
        return <RfpSubmitter />

      case 'MANAGE_RFP_TEMPLATES':
        return <RfpTemplateManager />

      default:
        return <div>ERROR: Unknown tab selected</div>
    }
  }

  componentWillUnmount () {
    this.props.clearRfpState()
  }
}

RfpStatus.propTypes = {
  tabs: PropTypes.array,
  selectedTabId: PropTypes.string
}

const mapStateToProps = state => ({
  tabs: state.optimization.rfp.tabs,
  selectedTabId: state.optimization.rfp.selectedTabId
})

const mapDispatchToProps = dispatch => ({
  setSelectedTabId: selectedTabId => dispatch(RfpActions.setSelectedTabId(selectedTabId)),
  clearRfpState: () => dispatch(RfpActions.clearRfpState())
})

const RfpStatusComponent = wrapComponentWithProvider(reduxStore, RfpStatus, mapStateToProps, mapDispatchToProps)
export default RfpStatusComponent
