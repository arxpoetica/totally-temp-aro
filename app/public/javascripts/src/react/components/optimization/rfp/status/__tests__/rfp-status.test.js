/* global test expect jest */
import React from 'react'
import { shallow } from 'enzyme'
import { RfpStatus } from '../rfp-status.jsx'

const tabs = [
  { id: 'LIST_PLANS', description: 'List all plans' },
  { id: 'SUBMIT_RFP', description: 'Submit RFP' },
  { id: 'MANAGE_RFP_TEMPLATES', description: 'Manage RFP templates' }
]

// -----------------------------------------------------------------------------
test('Default render and tab navigation', () => {
  const mockSetSelectedTabId = jest.fn()
  const component = shallow(
    <RfpStatus
      tabs={tabs}
      selectedTabId={tabs[0].id}
      setSelectedTabId={mockSetSelectedTabId}
    />
  )
  expect(component).toMatchSnapshot()

  component.find('#rfpStatusTab_SUBMIT_RFP').simulate('click')
  expect(component).toMatchSnapshot()
  expect(mockSetSelectedTabId).toHaveBeenCalledWith('SUBMIT_RFP')

  component.find('#rfpStatusTab_MANAGE_RFP_TEMPLATES').simulate('click')
  expect(component).toMatchSnapshot()
  expect(mockSetSelectedTabId).toHaveBeenCalledWith('MANAGE_RFP_TEMPLATES')

  component.find('#rfpStatusTab_LIST_PLANS').simulate('click')
  expect(component).toMatchSnapshot()
  expect(mockSetSelectedTabId).toHaveBeenCalledWith('LIST_PLANS')
})

// -----------------------------------------------------------------------------
test('RFP state cleared on unmount', () => {
  const mockClearRfpState = jest.fn()
  const component = shallow(
    <RfpStatus
      tabs={tabs}
      selectedTabId={tabs[0].id}
      clearRfpState={mockClearRfpState}
    />
  )
  component.unmount()
  expect(mockClearRfpState).toHaveBeenCalled()
})
