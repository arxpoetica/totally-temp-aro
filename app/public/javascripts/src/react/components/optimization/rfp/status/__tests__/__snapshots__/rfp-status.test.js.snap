// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default render and tab navigation 1`] = `
<div
  className="container pt-5 pb-5"
>
  <h2>
    RFP Plans
  </h2>
  <ul
    className="nav nav-tabs mb-3"
  >
    <li
      className="nav-item"
      key="LIST_PLANS"
    >
      <a
        className="nav-link active"
        href="#"
        id="rfpStatusTab_LIST_PLANS"
        onClick={[Function]}
      >
        List all plans
      </a>
    </li>
    <li
      className="nav-item"
      key="SUBMIT_RFP"
    >
      <a
        className="nav-link "
        href="#"
        id="rfpStatusTab_SUBMIT_RFP"
        onClick={[Function]}
      >
        Submit RFP
      </a>
    </li>
    <li
      className="nav-item"
      key="MANAGE_RFP_TEMPLATES"
    >
      <a
        className="nav-link "
        href="#"
        id="rfpStatusTab_MANAGE_RFP_TEMPLATES"
        onClick={[Function]}
      >
        Manage RFP templates
      </a>
    </li>
  </ul>
  <Component />
</div>
`;

exports[`Default render and tab navigation 2`] = `
<div
  className="container pt-5 pb-5"
>
  <h2>
    RFP Plans
  </h2>
  <ul
    className="nav nav-tabs mb-3"
  >
    <li
      className="nav-item"
      key="LIST_PLANS"
    >
      <a
        className="nav-link active"
        href="#"
        id="rfpStatusTab_LIST_PLANS"
        onClick={[Function]}
      >
        List all plans
      </a>
    </li>
    <li
      className="nav-item"
      key="SUBMIT_RFP"
    >
      <a
        className="nav-link "
        href="#"
        id="rfpStatusTab_SUBMIT_RFP"
        onClick={[Function]}
      >
        Submit RFP
      </a>
    </li>
    <li
      className="nav-item"
      key="MANAGE_RFP_TEMPLATES"
    >
      <a
        className="nav-link "
        href="#"
        id="rfpStatusTab_MANAGE_RFP_TEMPLATES"
        onClick={[Function]}
      >
        Manage RFP templates
      </a>
    </li>
  </ul>
  <Component />
</div>
`;

exports[`Default render and tab navigation 3`] = `
<div
  className="container pt-5 pb-5"
>
  <h2>
    RFP Plans
  </h2>
  <ul
    className="nav nav-tabs mb-3"
  >
    <li
      className="nav-item"
      key="LIST_PLANS"
    >
      <a
        className="nav-link active"
        href="#"
        id="rfpStatusTab_LIST_PLANS"
        onClick={[Function]}
      >
        List all plans
      </a>
    </li>
    <li
      className="nav-item"
      key="SUBMIT_RFP"
    >
      <a
        className="nav-link "
        href="#"
        id="rfpStatusTab_SUBMIT_RFP"
        onClick={[Function]}
      >
        Submit RFP
      </a>
    </li>
    <li
      className="nav-item"
      key="MANAGE_RFP_TEMPLATES"
    >
      <a
        className="nav-link "
        href="#"
        id="rfpStatusTab_MANAGE_RFP_TEMPLATES"
        onClick={[Function]}
      >
        Manage RFP templates
      </a>
    </li>
  </ul>
  <Component />
</div>
`;

exports[`Default render and tab navigation 4`] = `
<div
  className="container pt-5 pb-5"
>
  <h2>
    RFP Plans
  </h2>
  <ul
    className="nav nav-tabs mb-3"
  >
    <li
      className="nav-item"
      key="LIST_PLANS"
    >
      <a
        className="nav-link active"
        href="#"
        id="rfpStatusTab_LIST_PLANS"
        onClick={[Function]}
      >
        List all plans
      </a>
    </li>
    <li
      className="nav-item"
      key="SUBMIT_RFP"
    >
      <a
        className="nav-link "
        href="#"
        id="rfpStatusTab_SUBMIT_RFP"
        onClick={[Function]}
      >
        Submit RFP
      </a>
    </li>
    <li
      className="nav-item"
      key="MANAGE_RFP_TEMPLATES"
    >
      <a
        className="nav-link "
        href="#"
        id="rfpStatusTab_MANAGE_RFP_TEMPLATES"
        onClick={[Function]}
      >
        Manage RFP templates
      </a>
    </li>
  </ul>
  <Component />
</div>
`;
